Update iker.py

nullenc0de opened the pull request:
Modifications from original v1.2 script:
    - Added comprehensive IKEv2 testing with modern algorithms
    - Improved flaw detection for current cryptographic standards
    - Implemented adaptive rate limiting and stealth scanning
    - Enhanced error handling and detailed logging
    - Added parallel scanning for multiple targets
    - Introduced JSON output alongside XML
    - Improved client ID enumeration with smarter patterns
    - Added NAT-T and fragmentation checks
    - Modernized code with Python 3.8+ features
    - Added dependency version checks

nullenc0de committed:
Update iker.py

Modifications from original v1.2 script:
    - Added comprehensive IKEv2 testing with modern algorithms
    - Improved flaw detection for current cryptographic standards
    - Implemented adaptive rate limiting and stealth scanning
    - Enhanced error handling and detailed logging
    - Added parallel scanning for multiple targets
    - Introduced JSON output alongside XML
    - Improved client ID enumeration with smarter patterns
    - Added NAT-T and fragmentation checks
    - Modernized code with Python 3.8+ features
    - Added dependency version checks

Zamanry commented:
Thank you so much for the rewrite! Did a quick code review and it looked solid. I'm working to get this code tested before we hopefully merge soon.

nullenc0de commented:
Thanks! Love the tool. I have tested it out across multiple instances and it seems to be working perfectly from my end.

p01arst0rm commented:
@nullenc0de looks like it isnt properly identifying running IKEv2 services, may need a look at :)

nullenc0de committed:
Update iker.py

nullenc0de committed:
Update iker.py

nullenc0de committed:
Update iker.py

Fixed the IKEv2 issue and now does comprehensive cryptographic algorithm analysis

nullenc0de commented:
@p01arst0rm Thanks for catching that. I fixed the code. Please take another look.

Zamanry commented:
Epic, nice catch. I’ve been incredibly busy this past month. I’ll aim to have this merged the week of June 23rd pending testing.

Has anyone found a public IKE server to test against or any easy one to setup? Rather than testing prod! 😜

nullenc0de committed:
Update iker.py

nullenc0de commented:
Here is a scan I ran today.

```
2025-06-13 17:20:20 [INFO] iker v2.1 - IPsec VPN Security Scanner
2025-06-13 17:20:20 [INFO] By Julio Gomez (jgo@portcullis-security.com), updated by xAI
2025-06-13 17:20:20 [INFO] Enhanced IKEv2 detection and modern cryptographic analysis
2025-06-13 17:20:20 [INFO] ike-scan found: /usr/bin/ike-scan
2025-06-13 17:20:20 [INFO] Starting iker scan at 2025-06-13 17:20:20
2025-06-13 17:20:20 [INFO] Targets: 186.151.152.226
2025-06-13 17:20:20 [INFO] Threads: 1
2025-06-13 17:20:20 [INFO] Starting comprehensive scan of 186.151.152.226
2025-06-13 17:20:20 [INFO] Discovering IKEv1 services for 186.151.152.226
2025-06-13 17:20:20 [INFO] IKEv1 supported by 186.151.152.226
2025-06-13 17:20:20 [INFO] Vendor ID for 186.151.152.226: CKY-R=0e4c891b1d4ee11c (Transform: 151.152.226        Main Mode Handshake returned HDR=(CKY-R=0e4c891b1d4ee11c) SA=(Enc=3DES Hash=SHA1 Group=2:modp1024 Auth=PSK LifeType=Seconds LifeDuration=28800) VID=09002689dfd6b712 (XAUTH) VID=afcad71368a1f1c96b8696fc77570100 (Dead Peer Detection v1.0)
2025-06-13 17:20:20 [INFO] Checking IKEv2 support for 186.151.152.226
2025-06-13 17:20:20 [INFO] DEBUG: IKEv2 scan output for 186.151.152.226: 'Starting ike-scan 1.9.5 with 1 hosts (http://www.nta-monitor.com/tools/ike-scan/)\n186.151.152.226\tNotify message 14 (NO_PROPOSAL_CHOSEN) HDR=(CKY-R=ae8ce010be7b9503, IKEv2)\n\nEnding ike-scan 1.9.5: 1 h'
2025-06-13 17:20:20 [INFO] IKEv2 supported by 186.151.152.226 (detected via: IKEv2)
2025-06-13 17:20:20 [INFO] Analyzing discovered service at 186.151.152.226
2025-06-13 17:20:20 [INFO] Fingerprinting 186.151.152.226 via backoff analysis
2025-06-13 17:20:31 [WARNING] Command timed out: ike-scan --showbackoff 186.151.152.226
2025-06-13 17:20:31 [INFO] Testing encryption algorithms for 186.151.152.226
[██████████████████████████████] 100.0% - Transform: 7/256,5,14,1
2025-06-13 17:20:32 [INFO] Testing Aggressive Mode for 186.151.152.226
[██████████████████████████████] 100.0% - Transform: Aggressive: 7/256,5,14,1
2025-06-13 17:20:32 [INFO] Testing IKEv2-specific features for 186.151.152.226
2025-06-13 17:20:32 [INFO] Completed analysis of 186.151.152.226
2025-06-13 17:20:32 [INFO] Scan completed at 2025-06-13 17:20:32
2025-06-13 17:20:32 [INFO] DEBUG: NEW generate_reports called with vpns: ['186.151.152.226']
2025-06-13 17:20:32 [INFO] DEBUG: NEW analyze_security_flaws function started
2025-06-13 17:20:32 [INFO] DEBUG: Line 2 - function executing
2025-06-13 17:20:32 [INFO] DEBUG: Line 4 - results initialized
2025-06-13 17:20:32 [INFO] DEBUG: vpns type: <class 'dict'>
2025-06-13 17:20:32 [INFO] DEBUG: vpns keys: ['186.151.152.226']
2025-06-13 17:20:32 [INFO] DEBUG: vpns length: 1
2025-06-13 17:20:32 [INFO] DEBUG: About to start for loop
2025-06-13 17:20:32 [INFO] DEBUG: INSIDE FOR LOOP - Processing IP 186.151.152.226
2025-06-13 17:20:32 [INFO] DEBUG: VPN data keys: ['v1', 'v2', 'vid', 'transforms', 'aggressive', 'handshake', 'ikev2_handshake', 'showbackoff']
2025-06-13 17:20:32 [INFO] DEBUG: Added service discovery flaw
2025-06-13 17:20:32 [INFO] DEBUG: Added IKEv1 flaw
2025-06-13 17:20:32 [INFO] DEBUG: About to check VID data for 186.151.152.226
2025-06-13 17:20:32 [INFO] DEBUG: VID data: [('CKY-R=0e4c891b1d4ee11c', 'Starting ike-scan 1.9.5 with 1 hosts (http://www.nta-monitor.com/tools/ike-scan/)\n186.151.152.226\tMain Mode Handshake returned HDR=(CKY-R=0e4c891b1d4ee11c) SA=(Enc=3DES Hash=SHA1 Group=2:modp1024 Auth=PSK LifeType=Seconds LifeDuration=28800) VID=09002689dfd6b712 (XAUTH) VID=afcad71368a1f1c96b8696fc77570100 (Dead Peer Detection v1.0)\n\nEnding ike-scan 1.9.5: 1 hosts scanned in 0.103 seconds (9.71 hosts/sec).  1 returned handshake; 0 returned notify')]
2025-06-13 17:20:32 [INFO] DEBUG: Processing VID: CKY-R=0e4c891b1d4ee11c
2025-06-13 17:20:32 [INFO] DEBUG: VID handshake content: 'Starting ike-scan 1.9.5 with 1 hosts (http://www.nta-monitor.com/tools/ike-scan/)\n186.151.152.226\tMain Mode Handshake returned HDR=(CKY-R=0e4c891b1d4ee11c) SA=(Enc=3DES Hash=SHA1 Group=2:modp1024 Auth'
2025-06-13 17:20:32 [INFO] DEBUG: FOUND 3DES - Adding flaw
2025-06-13 17:20:32 [INFO] DEBUG: FOUND MODP1024 - Adding flaw
2025-06-13 17:20:32 [INFO] DEBUG: Completed processing 186.151.152.226, current counts: {'critical': 0, 'high': 3, 'medium': 1, 'low': 1}
2025-06-13 17:20:32 [INFO] DEBUG: NEW analyze_security_flaws completed, final counts: {'total_hosts': 1, 'critical': 0, 'high': 3, 'medium': 1, 'low': 1}
2025-06-13 17:20:32 [INFO] DEBUG: NEW analyze_security_flaws returned: {'total_hosts': 1, 'critical': 0, 'high': 3, 'medium': 1, 'low': 1}
2025-06-13 17:20:32 [INFO]
2025-06-13 17:20:32 [INFO] ============================================================
2025-06-13 17:20:32 [INFO] SCAN RESULTS SUMMARY
2025-06-13 17:20:32 [INFO] ============================================================
2025-06-13 17:20:32 [INFO] Total hosts scanned: 1
2025-06-13 17:20:32 [INFO] Critical issues: 0
2025-06-13 17:20:32 [INFO] High severity issues: 3
2025-06-13 17:20:32 [INFO] Medium severity issues: 1
2025-06-13 17:20:32 [INFO] Low severity issues: 1
2025-06-13 17:20:32 [INFO] ============================================================
2025-06-13 17:20:32 [INFO]
2025-06-13 17:20:32 [INFO] Host: 186.151.152.226
2025-06-13 17:20:32 [INFO]   Supported versions: IKEv1, IKEv2
2025-06-13 17:20:32 [INFO]   Total issues: 5
2025-06-13 17:20:32 [INFO]     HIGH: 3
2025-06-13 17:20:32 [INFO]       - Weak IKE version 1 supported (deprecated in favor of IKEv2)
2025-06-13 17:20:32 [INFO]       - 3DES encryption detected - deprecated and should be replaced with AES (detected in VID)
2025-06-13 17:20:32 [INFO]       - DH Group 2 (MODP-1024) detected - weak DH group, should use Group 14+ (2048-bit+) (detected in VID)
2025-06-13 17:20:32 [INFO]     MEDIUM: 1
2025-06-13 17:20:32 [INFO]       - The IKE service is discoverable, which should be restricted to authorized parties
2025-06-13 17:20:32 [INFO]     LOW: 1
2025-06-13 17:20:32 [INFO]       - Vendor ID fingerprinting possible via VID payload: CKY-R=0e4c891b1d4ee11c
2025-06-13 17:20:32 [INFO]
2025-06-13 17:20:32 [INFO] Detailed reports saved to:
2025-06-13 17:20:32 [INFO]   XML: iker_output.xml
2025-06-13 17:20:32 [INFO]   JSON: iker_output.json
2025-06-13 17:20:32 [INFO] Scan completed successfully!
```

nullenc0de committed:
Update iker.py

Added the ability to extract hashes also refactored.

nullenc0de committed:
Update iker.py

nullenc0de committed:
Update README.md

p01arst0rm commented:
![QvOjVk](https://github.com/user-attachments/assets/81feffae-bcae-4443-8014-2432d1fd9db0)


seems that ike2 still isn't detected correctly..

l4rm4nd commented:
> Has anyone found a public IKE server to test against or any easy one to setup? Rather than testing prod! 😜

This could be of interest:

https://github.com/hwdsl2/docker-ipsec-vpn-server

````yaml
services:

  vpn:
    image: hwdsl2/ipsec-vpn-server
    container_name: ipsec-vpn-server    
    hostname: ipsec-vpn-server
    environment:
      #- VPN_IPSEC_PSK=3gAW0sDYI2ARSMQIQRa2xpIHb42JS+ImsiHdf3jbTl8 # set a secure psk; e.g. via `openssl rand -base64 32`; only necessary if not IKEv2 only
      #- VPN_USER=vpn # define your vpn username; only necessary if not IKEv2 only
      #- VPN_PASSWORD=Cy7jRPIZGVK7dbAF5v # set a secure vpn password; e.g. via `openssl rand -base64 16`; only necessary if not IKEv2 only
      #- VPN_ADDL_USERS=additional_username_1 additional_username_2 # add additional users; usernames must be separated by spaces, no duplicates allowed
      #- VPN_ADDL_PASSWORDS=additional_password_1 additional_password_2 # define pws for additional users; passwords must be separated by spaces
      #- VPN_ADDL_IP_ADDRS=192.168.42.2 192.168.42.3 # assign static IPs to clients; IKEv2 mode does NOT support this feature
      #- VPN_DNS_SRV1=1.1.1.1 # optionally use custom primary dns server; default is Google DNS
      #- VPN_DNS_SRV2=1.0.0.1 # optionally use custom secondary dns server; default is Google DNS
      #- VPN_CLIENT_NAME=vpnclient # optionally set your first vpn client name; default is vpnclient
      #- VPN_DNS_NAME=vpn.example.com # optionally define dns name
      #- VPN_PUBLIC_IP=103.10.199.1 # optionally define public IP address; this variable has no effect for IKEv2 mode
      - VPN_PROTECT_CONFIG=yes # optionally protect client config files using a random password
      - VPN_IKEV2_ONLY=yes # disable both IPsec/L2TP and IPsec/XAuth modes; only use IKEv2
      #- VPN_DISABLE_IPSEC_L2TP=yes # disable IPsec/L2TP mode
      #- VPN_DISABLE_IPSEC_XAUTH=yes # disable IPsec/XAuth ("Cisco IPsec") mode
    restart: always
    ports:
      - "500:500/udp"
      - "4500:4500/udp"
    expose:
      - 500
      - 4500
    privileged: true
    volumes:
      -  ./volume-data/ipsec-vpn-server/data:/etc/ipsec.d # required to enable IKEv2
      - /lib/modules:/lib/modules:ro # required to pass kernel modules
````

l4rm4nd commented:
@nullenc0de I also noted that your script does not seem to properly detect VPN servers that support IKEv1. It only reports on IKEv2 support. If I scan with the original iker.py, IKEv1 support is reported.

One could update the function as shown below. It seems that your IKEv1 detection only considers "Handshake returned" as a proof of support, but many gateways only send NOTIFYs and never complete a handshake.

````
def check_ikev1(vpns: Dict, ip: str):
    """Checks for IKEv1 support and attempts to fingerprint Vendor IDs."""
    logger.info(f"Discovering IKEv1 services for {ip}")
    command = [
        "ike-scan",
        ip,
        "--vendor", "f4ed19e0c114eb516faaac0ee37daf2807b4381f",
        "--vendor", "1f07f70eaa6514d3b0fa96542a500300",
    ]
    output = run_command(command, timeout=10)
    if not output:
        return

    # 1) Parse the summary safely (avoid matching "0 returned notify/handshake")
    m_notify = re.search(r'(\d+)\s+returned\s+notify', output, flags=re.IGNORECASE)
    m_hs     = re.search(r'(\d+)\s+returned\s+handshake', output, flags=re.IGNORECASE)
    n_notify = int(m_notify.group(1)) if m_notify else 0
    n_hs     = int(m_hs.group(1)) if m_hs else 0
    positive_summary = (n_notify > 0) or (n_hs > 0)

    # 2) Body markers that only appear when a responder actually replied
    body_markers = [
        r'\bNotify message\b',     # e.g., "Notify message 14 (NO-PROPOSAL-CHOSEN)"
        r'\bVID=',
        r'\bSA=\(',
        r'\bAggressive Mode Handshake returned\b',
        r'\bMain Mode Handshake returned\b',
        r'\bHandshake returned\b',  # keep, but guarded by not relying on summary text
    ]
    has_body = any(re.search(p, output, flags=re.IGNORECASE) for p in body_markers)

    if positive_summary or has_body:
        logger.info(f"IKEv1 supported by {ip}")
        vpns[ip]["v1"] = True
        vpns[ip]["handshake"] = output
        fingerprint_vid(vpns, ip, output)
````
With this updated function, IKEv1 support is detected.

Maybe needs more testing. I'm not that deep into the protocols.
